coveo_stew/ci/runner.py

```diff
@@ -119,3 +137,141 @@ def _output_generic_report(self, environment: PythonEnvironment) -> None:
+    def autofix_checks(self) -> List[ContinuousIntegrationRunner]:
+        """
+        Autofix checks receive a special treatment since they will potentially change line numbers,
+        making reports inaccurate.
+        """
+        return [check for check in self.checks if check.supports_auto_fix]
+
+    @cached_property
+    def non_autofix_checks(self) -> List[ContinuousIntegrationRunner]:
+        return [check for check in self.checks if check not in self.autofix_checks]
+
+    async def orchestrate(self, auto_fix: bool = False) -> None:
+        """Orchestrates this CIPlan by launching runners in the correct order."""
+        runs: List[Run] = []
+        echo.step(
+            f"Planned {len(self.checks)} runners for {self.environment.pretty_python_version}"
+        )
+
+        if auto_fix:
+            # run autofix first, in parallel
+            runs.append(run := Run(self.environment, self.autofix_checks))
+            await run.run_and_report(parallel=self.parallel)
+
+            if run.overall_status is RunnerStatus.CheckFailed:
+                await run.run_and_report(auto_fix=True, feedback=False, parallel=False)
+                await run.run_and_report(parallel=self.parallel)  # verify that autofix worked
+
+            # run all other runners
+            runs.append(run := Run(self.environment, self.non_autofix_checks))
+            await run.run_and_report(parallel=self.parallel)
+
+        else:
+            runs.append(run := Run(self.environment, self.checks))
+            await run.run_and_report(parallel=self.parallel)
+
+        overall_status = get_overall_run_status(*runs)
+        echo.success(
+            f"The CI run for {self.environment.pretty_python_version} completed with status: {overall_status}"
+        )
+
+        if exceptions := [exception for run in runs for exception in run.exceptions]:
+            raise ExitWithFailure(
+                suggestions=(
+                    "If a command should be treated as a check failure, specify `check-failed-exit-codes`",
+                    "Reference: https://github.com/coveo/stew/blob/main/README.md#options)",
+                    "Try the commands in a shell to troubleshoot them faster.",
+                ),
+                failures=(
+                    f"\n------- [{runner} failed unexpectedly] -------\n\n{str(ex)}\n"
+                    for runner, ex in exceptions
+                ),
+            ) from CheckError("Unexpected errors occurred when launching external processes.")
+
+
+@dataclass
+class Run:
+    """The Run is a stateful object that runs checks and reports the results to the user."""
+
+    environment: PythonEnvironment
+    checks: Sequence[ContinuousIntegrationRunner]
+
+    @cached_property
+    def exceptions(self) -> List[Tuple[ContinuousIntegrationRunner, DetailedCalledProcessError]]:
+        """Exceptions are stored here after the run. Exceptions are cleared when `run_and_report` is called."""
+        return []
```

Comment:
What happens if we call this too early, would the cache always return `[]`? Could it even happen?